File: services/crm-service/src/main/java/com/crm/crm_backend/model/entity/Invoice.java
package com.crm.crm_backend.model.entity;


import com.crm.crm_backend.model.enums.InvoiceStatus;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "invoices")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Invoice {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "invoice_number", nullable = false, unique = true)
    private String invoiceNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "sales_order_id")
    private SalesOrder salesOrder;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InvoiceStatus status;

    @Column(name = "invoice_date")
    private LocalDate invoiceDate;

    @Column(name = "due_date")
    private LocalDate dueDate;

    @Column(precision = 15, scale = 2)
    private BigDecimal subtotal;

    @Column(precision = 15, scale = 2)
    private BigDecimal taxAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal discountAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal totalAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal paidAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal balanceAmount;

    private String currency;

    @Column(length = 2000)
    private String notes;

    @Column(name = "last_reminder_sent_at")
    private LocalDateTime lastReminderSentAt;

    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @PrePersist
    public void prePersist() {

        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();

        if (status == null) {
            status = InvoiceStatus.DRAFT;
        }

        if (currency == null) {
            currency = "ETB";
        }

        if (paidAmount == null) {
            paidAmount = BigDecimal.ZERO;
        }

        if (balanceAmount == null && totalAmount != null) {
            balanceAmount = totalAmount;
        }
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/features/crm/types/invoice.ts
export type InvoiceStatus =
  | "DRAFT"
  | "PENDING"
  | "SENT"
  | "PAID"
  | "OVERDUE"
  | "CANCELLED";

export type Invoice = {
  id: number;
  invoiceNumber: string;
  salesOrderId: number | null;
  customerId: number;
  status: InvoiceStatus;
  invoiceDate: string | null;
  dueDate: string | null;
  subtotal: number | null;
  taxAmount: number | null;
  discountAmount: number | null;
  totalAmount: number | null;
  paidAmount: number | null;
  balanceAmount: number | null;
  currency: string | null;
  notes: string | null;
  createdAt: string | null;
  updatedAt: string | null;
};

export type InvoiceCreateInput = {
  salesOrderId: number;
  customerId: number;
  invoiceDate?: string;
  dueDate?: string;
  taxAmount?: number;
  discountAmount?: number;
  currency?: string;
  notes?: string;
};

export type InvoiceUpdateInput = {
  status?: InvoiceStatus;
  dueDate?: string;
  paidAmount?: number;
  notes?: string;
};
